### PR TITLE
Implement todict and fromdict for ASE calculator

### PR DESCRIPTION
**Pull request by Luthaf**
Part of #430.

We added an ASE calculator for metatensor models in #405, leaving `fromdict` and `todict` unimplemented, this PR tries to add these functions (which are standard and required in ASE as far as I understand).

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [ ] ~Documentation updated (for new features)?~
 - [X] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
📚 Documentation preview 📚: https://metatensor--472.org.readthedocs.build/en/472/

### CODE REVIEW

File: python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py [hunk 1]
```diff
@@ -88,12 +90,23 @@ def __init__(
 
     def todict(self):
         # used by ASE to save the calculator
-        raise NotImplementedError("todict is not yet implemented")
+        try:
+            model = torch.jit.script(self._model)
+        except RuntimeError as e:
+            raise RuntimeError("could not convert the module to TorchScript") from e
```
Comment (sirmarcel) [inline python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py:96]: I'm unfamiliar with this syntax, what does this line do with `e`?

**Review by sirmarcel — COMMENTED** [review]
As far as I understand this code, which is only medium given my limited understanding of `metatensor` and `torchscript`, this is fine technically.

My only concern is more philosophical: If we consider that `metatensor` will probably at some point deal with models that are "thick" in the sense of having millions of weights, is it a good idea to have an automated facility that dumps the serialised version of these weights into every timestep of a `.traj`? 

As far as I know, `ase` expects `todict` to return something like the parameters of an electronic structure code, rather than a full serialisation of the underlying calculator. So I think the more `ase`-style approach would be to return a dict with the path to the serialised model, rather than the full serialised model. Of course, at the cost of portability... 

So, overall, this is fine to merge in principle, but I would be curious what the intended use case for this is, and see whether this can be better served in some other way.

File: python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py (hunk 1 above)
Comment (Luthaf) [reply python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py:96]: This creates a new exception with a `cause` field, which is then displayed to the user:

```py
try:
    raise Exception("first one")
except Exception as e:
    raise Exception("second one") from e
```

```
Traceback (most recent call last):
  File "tmp.py", line 2, in <module>
    raise Exception("first one")
Exception: first one

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "tmp.py", line 4, in <module>
    raise Exception("second one") from e
Exception: second one
```

**Comment by Luthaf** [comment]
I did not realize this was dumped at every timestep. Agreed that this is not ideal. If the user gave us the path to a model, saving the path again makes sense, not sure how we should handle the case where the user gives a large in-memory model.

File: python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py (hunk 1 above)
Comment (sirmarcel) [reply python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py:96]: Okay -- thanks.

**Comment by sirmarcel** [comment]
Yeah, it looks like this gets called at every timestep: https://gitlab.com/ase/ase/-/blob/master/ase/io/trajectory.py#LC174. I don't think this gets used by particularly serious MD codes, so maybe this is completely fine, but it should probably at least be documented in case someone wonders why their `.traj` files are extremely huge...

**Comment by sirmarcel** [comment]
Okay, I saw your comment now. As far as I know, this is not a required interface for `ase`.  The only thing you need to have is a `parameters` dict (but it doesn't matter particularly what's in there). So one might want to write something informative there, like where we found the model that made up the calculator, but unless we have a pressing need to implement the `todict`/`fromdict` interface, I personally wouldn't bother keeping it around and having it potentially cause issues.

**Comment by Luthaf** [comment]
I've updated the code to store the model path in `todict`, which should solve the size issue. The error message should also point users who need this to work to save the model in a file themself. 

> As far as I know, this is not a required interface for ase.

It is part of the interface and used by some of the code in ASE =/. I agree that we should not be implementing everything just for fun, but the idea is also that this project should be usable by people outside the lab, who might not take the time to report issues if something is broken for them. It seems to me that this code is simple enough to not add too much maintenance work in the future =)

**Comment by sirmarcel** [comment]
Sounds good -- what I meant is that `ase` (as far as I know, as discussed in Slack) doesn't expect these functions to do anything very well-defined, and I think in principle one can get away with only `todict` and not `fromdict`. I think this is a good compromise, we can always adjust further if user needs arise. Once the pipeline is done, this can be merged. Thank you!

**Comment by Luthaf** [comment]
codecov is not behaving today ...

**Pull request merged**